### DISCUSSION TITLE
transcoding for mp2 files

### DISCUSSION THREAD

**Discussion by nobanzai** [gerbera/gerbera#3305] [General]
I want to configure Gerbera, so that mp2 files are transcoded to mp3 for a few clients, that miss the capability to play them, but not for others, that have this capability.
I already got this working for 24bit vs. 16bit FLACs, but because mp2 and mp3 are both audio/mpeg files, I'm not sure, if I can configure them the same way.
Just to make sure, I do understand the complex config settings, can anyone check my config snippet below and tell me, if it is correct or if I can achieve this goal by simpler means/lesser config parms:

```
...
  <clients
    enabled="yes"
    cache-threshold="8760"
    bookmark-offset="10"
  >
    <client
      userAgent="Allegro-Software-RomPlug"
      flags="TRANSCODE1"
    />
    <client
      userAgent="Allegro-Software-WebClient"
      flags="TRANSCODE1"
    />
...
  <import>
    ...
    <mappings>
      <extension-mimetype ignore-unknown="no" case-sensitive="no">
        ...
        <map from="mp2" to="audio/x-mp2"/>
        <map from="mp3" to="audio/mpeg"/>
      </extension-mimetype>
      ...
      <mimetype-contenttype>
        ...
        <treat mimetype="audio/mpeg" as="mp3"/>
        <treat mimetype="audio/x-mp2" as="mp2"/>
      </mimetype-contenttype>
      ...
      <contenttype-dlnaprofile>
        ...
        <map from="mp2" audioCodec="mp2" to="MP2_MPS" />
      </contenttype-dlnaprofile>
    </mappings>
  ...
  </import>
  ...
  <transcoding enabled="yes">
    <mimetype-profile-mappings allow-unused="yes">
      <transcode
        mimetype="audio/x-mp2"
        source-profile="MP2_MPS"
        client-flags="TRANSCODE1"
        using="mh-mp2mp3"
      />
      ...
    </mimetype-profile-mappings>
    <profiles allow-unused="yes">
      <profile
        name="mh-mp2mp3"
        enabled="yes"
        type="external"
      >
        <mimetype>audio/mpeg</mimetype>
        <dlna-profile>MP3</dlna-profile>
        <accept-url>no</accept-url>
        <first-resource>yes</first-resource>
        <hide-original-resource>no</hide-original-resource>
        <accept-ogg-theora>no</accept-ogg-theora>
        <agent
          command="ffmpeg"
          arguments="-hide_banner -i %in -f mp3 -y %out"
        />
        <buffer size="1048576" chunk-size="131072" fill-size="262144"/>
      </profile>
    </profiles>
  </transcoding>
...

```
TIA!

**Comment by udippel** [comment]
Stupid question(s), and depending on the clients, though:
`<hide-original-resource>no</hide-original-resource>`

When you just transcode all and any, your stream contains all and any as mp2 and mp3. What's wrong about that? Except some of your clients can't handle that situation. Maybe I was lucky here with my ogg files, that play perfectly well on a SONY with 'first'?

Or, what makes you unhappy with straightforward transcoding all? I guess, we talk about listening, not editing, on the clients?

**Reply by nobanzai** [reply]
I have renderers, that are able to play mp3 and mp2 (Majik DSM) and I have renderers without the capability to play mp2 (Streamiums).

So I want to transcode mp2 to mp3 for the Streamiums and  would like to get the original file streamed to the Majik DSM.

What exactly is stupid with this?

Besides that, I have a control point (upplay), that is able to stream depending on the capabilites of the renderers.

**Reply by udippel** [reply]
Sorry, didn't mean YOUR question, but mine! 
I do understand that you want this; and yet MY (maybe stupid) question: Why; if the two formats could be played by any client? Respectively, since the audio quality is principally identical, what's wrong with either?

My apologies, though, once again, for the completely mis-formulated question!

**Reply by nobanzai** [reply]
All good, no worries!

Could I explain at least, why I want this config?

Its the same reason as for the 24bit vs. 16bit FLACs. One renderer is capable to play them, the other ones not.

**Reply by udippel** [reply]
Thanks for accepting my apology!

It seems we turn in circles, though. 
Is it your intention, kind of academic exercise, to define a matrix of formats and clients and arbitrarily define which client is supplied with which format?
Or is it just your intention to play all your files on any client? 

From that earlier thread I was under the impression of the former. Listening to mp3, FLAC_16 or FLAC_24 wouldn't make much of a difference. At least on my account. 

From what I read above in this thread, I thought you wanted the latter. Therefore I added my example of having a huge collection of ogg, that not all of my clients can handle. Therefore, I transcode all audio into mp3. So my clients are fed with mp3, and the original. So far, all have been playing whatever I threw at them.

To sidetrack this matter: While this works perfectly okay on my tiny ex-thin-client server with audio, it doesn't at all with video, due to the - in comparison - huge demand on processing power. And with many different video formats. 
Therefore I bought an Android 13 X98K Smart TV Box for 29,95 € including shipping. Installed VLC on it, and any trouble is a trouble of the past.

**Reply by nobanzai** [reply]
> Thanks for accepting my apology!
> 
> It seems we turn in circles, though. Is it your intention, kind of academic exercise, to define a matrix of formats and clients and arbitrarily define which client is supplied with which format? Or is it just your intention to play all your files on any client?

Uhm, the main intention is to play all songs on all devices. And I want to avoid transcoding, if it is not necessary. And - believe it or not - you can hear the difference between a 24bit FLAC and an MP3 on the Majik DSM, while there are no differences between 16bit FLAC and MP3 on the Streamiums.
But I also tend to do things as perfect as possible and I always want to understand everything I use in total.
So both of your mentioned alternatives are at least partially true.

**Reply by udippel** [reply]
Oh, that clears things up.
I have no qualms about your intention of doing things perfectly. It would actually be a great progress to have this 'controlable transcoding'. But the method of using a linear set of exclusions, inclusions, exceptions, formats and IPs won't get one very far. It would actually be less than perfect, structurally. 
The first method that comes into mind would really be a matrix - or matrices - where input formats are tagged/ticked with transcoders (or just not) and sent to, respectively processed depending on the clients' abilities. 
(Another time: so sad I'm retired and have no more students to pass this on to as project.)

**Reply by nobanzai** [reply]
> Oh, that clears things up. I have no qualms about your intention of doing things perfectly. It would actually be a great progress to have this 'controlable transcoding'. But the method of using a linear set of exclusions, inclusions, exceptions, formats and IPs won't get one very far. It would actually be less than perfect, structurally. The first method that comes into mind would really be a matrix - or matrices - where input formats are tagged/ticked with transcoders (or just not) and sent to, respectively processed depending on the clients' abilities. (Another time: so sad I'm retired and have no more students to pass this on to as project.)

At the moment, this is the problem: The control point knows the caps of the renderers, Gerbera does not know them, because it never sees them, if they don't use their integrated control point as for example the Streamiums. So Gerbera can only provide alternatives, the control point has to chose from depending on the caps of the renderers. And if we have 24bit and 16bit FLACs, the control point isn't able to distinguish, because it only knows, that the renderer is able to play audio/x-flac, not what bit depth exactly. Similar for MP2 and MP3, because both are of mime type audio/mpeg.

So the control point has to learn a few new settings, one of them is: Ignore the caps of a renderer, another could be: Renderer can't play MP2. I had long discussions with the developer of upplay because of this, and he has the same opinion.

And regarding the results: At the moment I am not very far from my final goal:
Renderers get 24bit FLAC, if they can handle them, or get them transcoded to MP3, if they can't handle them.
All renderers get MP2 files transcoded to MP3.

Two things are missing:
1. Transcode 24bit FLAC to 16bit FLAC instead of MP3 and
2. don't transcode MP2 to MP3, if the renderer is able to play MP2.

**Reply by udippel** [reply]
Okay, it is much clearer now what you actually want. 
1. Is most obvious, I guess.
2. I still don't see, why transcoding to MP3 with  `<first-resource>no</first-resource>` shouldn't play mp2 even on clients with little 'intelligence', but mp2-ability. At least I for one wouldn't see any imperfections if a second stream - unused - accompanies the desired stream as played.

**Reply by nobanzai** [reply]
>     2. I still don't see, why transcoding to MP3 with  `<first-resource>no</first-resource>` shouldn't play mp2 even on clients with little 'intelligence', but mp2-ability. At least I for one wouldn't see any imperfections if a second stream - unused - accompanies the desired stream as played.

Because nothing knows, that this renderer is able to play MP2 - Gerbera not, because it doesn't "see" the renderer at all, upplay because it can't distinguish between MP2 and MP3, because both are basically of mime type audio/mpeg.

upplay will learn to distinguish between DLNA profiles in the near future, then my second goal can probably achieved.

**Reply by nobanzai** [reply]
Btw. with `<first-resource>no</first-resource>` the first resource is the MP2 one, which leads to a very strange error message on the Streamiums "server is not available" 8-<

**Reply by udippel** [reply]
Exactly here is where I wanted to be some time earlier. It doesn't work for the FLAC24->FLAC16, though.
But for the rest it is supposed to work in principle: You stream the original, first, stream, whichever it is, and if a client can't handle this one, it will (hopefully) use the fall-back stream of mp3.

**Reply by nobanzai** [reply]
If I've understood correctly, that is only possible at the moment, if the renderer directly talks to the media server.
And indeed it works, if the Streamium directly talks to Gerbera via it's integrated control point.

**Reply by udippel** [reply]
Okay, now I am satisfied, so to say. That's the end of my knowledge. 
Sorry for not being able to be more helpful once again. I simply jump in, when things defy my 'internal logic'. And now your two points as above are clear to me. I hope you can forgive me once again!

**Reply by nobanzai** [reply]
Can only repeat "All good, no worries" :)
